File: Python/Trie.py
class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.item = ""

root = TrieNode()

def addWord(word):
    node = root
    for c in word:
        index = ord(c) - ord('a')
        if not node.children[index]:
            node.children[index] = TrieNode()
        node = node.children[index]
    node.item = word

def search(word):
    return match(list(word), 0, root)

def match(ch_list, k, node):
    if k == len(ch_list):
        return bool(node.item) # return True if the node is not empty
    if ch_list[k] != '.':
        index = ord(ch_list[k]) - ord('a')
        return node.children[index] and match(ch_list, k + 1, node.children[index])
    else:
        for child in node.children:
            if child and match(ch_list, k + 1, child):
                return True
    return False

addWord("word")
addWord("hello")
print(search("word") == True)
print(search("hello") == True)
print(search("abc") == False)

File: Java/Trie.java
import java.util.HashMap;
import java.util.Map;

public class Trie {
    public class TrieNode {
        public Map<Character, TrieNode> children = new HashMap<Character, TrieNode>(); 
        public boolean isWord = false;
    }
    
    private TrieNode root = new TrieNode();

    public void addWord(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            node = node.children.computeIfAbsent(c, k -> new TrieNode());
        }
        node.isWord = true;
    }

    public boolean search(String word) {
        return match(word.toCharArray(), 0, root);
    }
    
    private boolean match(char[] chs, int k, TrieNode node) {
        if (k == chs.length) return node.isWord;   
        if (chs[k] != '.') {
            return node.children.containsKey(chs[k]) && match(chs, k + 1, node.children.get(chs[k]));
        } else {
            for (Map.Entry<Character, TrieNode> entry : node.children.entrySet()) {
                if (match(chs, k + 1, entry.getValue())) {
                    return true;
                }
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Trie t = new Trie();
        t.addWord("word");
        t.addWord("hello");
        System.out.println(t.search("wo..") == true);
        System.out.println(t.search("word") == true);
        System.out.println(t.search("hello") == true);
        System.out.println(t.search("abc") == false);
    }
}
